tests/test_trt_llm.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import pytest
+import torch
+
+from nemo_aligner.utils import parallel_state
+from nemo_aligner.utils.trt_llm import GPTGenerateTRTLLM
+
+
+@pytest.mark.mpi
+def test_trtllm_does_not_insert_padding(dummy_actor_gpt_model_with_pp):
+    trtllm_generate = GPTGenerateTRTLLM(
+        model_cfg=dummy_actor_gpt_model_with_pp.cfg,
+        end_strings=["<extra_id_1>"],
+        tokenizer=dummy_actor_gpt_model_with_pp.tokenizer,
+        max_input_len=dummy_actor_gpt_model_with_pp.cfg.encoder_seq_length // 2,
+        max_generation_length=dummy_actor_gpt_model_with_pp.cfg.encoder_seq_length // 2,
+    )
+    trtllm_generate.refit(dummy_actor_gpt_model_with_pp)
+
+    batch_size = 4
+    max_seq_len = dummy_actor_gpt_model_with_pp.cfg.encoder_seq_length
+    prompt_tokens = torch.ones((batch_size, max_seq_len), dtype=torch.int32)
+    prompt_lengths = torch.tensor([10, 20, 30, 40])
+
+    output_ids, response_lengths = trtllm_generate._generate([prompt_tokens, prompt_lengths])
+    max_length = response_lengths.max().item()
+
+    # TRTLLM with PP has sometimes erroneously inserts padding:
+    # As an example when we have the input:
+    #     [[prompt tok, PAD, PAD], [prompt tok, prompt tok, prompt tok]]
+    # The output when PP is enabled becomes:
+    #     [[prompt tok, PAD, PAD, resp_tok, resp_tok], [prompt tok, prompt tok, prompt tok, resp_tok, resp_tok]]
+    # Therefore we need this logic to get rid of the padding in the middle of the tensor.
+    # Furthermore, TRTLLM only produces valid outputs on the source rank, so we can only process it here
+    # and rely on the aligner broadcast to get it to the other ranks. Miraculously, the length
+    # is still correct on the non src ranks
+    if (
```

Comment:
do we still need this given we've removed it from the main trt-llm file?

Reply:
So my reasons for keeping it are mainly:
1. I don't know what PR fixed this or if this is input dependent, so I thought to leave it for posterity
2. this unit test is also an exemplar of how to test TRTLLM stuff with mpi